controller/ControllerRTL.py

```diff
@@ -0,0 +1,94 @@
+"""
+==========================================================================
+ControllerRTL.py
+==========================================================================
+Simple controller for CGRA.
+
+Author : Cheng Tan
+  Date : Dec 2, 2024
+"""
+
+
+from pymtl3 import *
+from pymtl3.stdlib.primitive import RegisterFile
+from ..lib.basic.en_rdy.ifcs import SendIfcRTL, RecvIfcRTL
+from ..lib.basic.val_rdy.ifcs import SendIfcRTL as ValRdySendIfcRTL
+from ..lib.basic.val_rdy.ifcs import RecvIfcRTL as ValRdyRecvIfcRTL
+from ..lib.opt_type import *
+
+
+class ControllerRTL(Component):
+
+  def construct(s, RingPktType, CGRADataType, CGRAAddrType):
+
+    # # Constant
+
+    # AddrType = mk_bits( clog2( data_mem_size ) )
+
+    # Interface
+
+    # # Read/write Data from local CGRA data memory.
+    # s.recv_raddr_from_local = RecvIfcRTL(AddrType)
+    # s.send_rdata_from_local = SendIfcRTL(DataType)
+    # s.recv_waddr_from_local = RecvIfcRTL(AddrType)
+    # s.recv_wdata_from_local = RecvIfcRTL(DataType)
+
+    # # Read/write Data to local CGRA data memory.
+    # s.recv_raddr_to_local = RecvIfcRTL(CGRAAddrType)
+    # s.send_rdata_to_local = SendIfcRTL(CGRADataType)
+    # s.recv_waddr_to_local = RecvIfcRTL(CGRAAddrType)
+    # s.recv_wdata_to_local = RecvIfcRTL(CGRADataType)
+
+    # Request from other CGRA.
+    s.recv_from_other = ValRdyRecvIfcRTL(RingPktType)
+    s.send_to_other = ValRdySendIfcRTL(RingPktType)
+
+    # Request from master.
+    s.recv_from_master = RecvIfcRTL(CGRADataType)
+    s.send_to_master = SendIfcRTL(CGRADataType)
+
+
+    # # TODO: below ifcs should be connected through another NoC within
+    # # one CGRA, instead of per-tile and performing like a bus.
+    # # Configuration signals to be written into and read from per-tile
+    # # control memory.
+    # s.recv_waddr = [RecvIfcRTL(AddrType) for _ in range(s.num_tiles)]
+    # s.recv_wopt = [RecvIfcRTL(CtrlType) for _ in range(s.num_tiles)]
+
+    # s.send_waddr = [SendIfcRTL(AddrType) for _ in range(s.num_tiles)]
+    # s.send_wopt = [SendIfcRTL(CtrlType) for _ in range(s.num_tiles)]
+
+    # # Cmd to invoke/terminate tiles execution.
+    # s.recv_cmd = [RecvIfcRTL(b2) for _ in range(s.num_tiles)]
+    # s.send_cmd = [SendIfcRTL(b2) for _ in range(s.num_tiles)]
+
+
+    s.pkt2data = Wire(CGRADataType)
+    s.data2pkt = Wire(RingPktType)
+
+    @update
+    def update_data():
+      s.pkt2data.payload @= s.recv_from_other.msg.payload
+      s.data2pkt.payload @= s.recv_from_master.msg.payload
+
+
+    # Can also be @update instead of @update_ff
+    # @update
+    @update_ff
+    def update_controller():
+      s.recv_from_other.rdy <<= s.send_to_master.rdy
+      s.send_to_master.en <<= s.recv_from_other.val & s.send_to_master.rdy
+      s.send_to_master.msg <<= s.pkt2data
+
+      s.recv_from_master.rdy <<= s.send_to_other.rdy
+      s.send_to_other.val <<= s.recv_from_master.en

```

Comment:
should it be `s.send_to_other.val <<= s.recv_from_master.en & s.send_to_other.rdy` ?

Reply:
`val` means there is data pending to be sent out. `en` is a proactive action that `val` shouldn't depend on it. This line causes hanging, i.e., `val` never be `true`.

I change this by including a `queue` in the latest commit, and change this to `val = queue.count > 0`.